src/lua/circllhist.lua

```diff
@@ -0,0 +1,565 @@
+    -- C: min, max < 0
+    if (0 <= f_min) then -- Case A
+      local f_hi = self:quantiles(0.9)
+      if (f_max < 4*f_md) then
+        -- max is not unreasonably high: use it
+        return 0, f_max
+      else
+        -- put f_hi at 75% of y-scale
+        return 0, f_hi / 0.75
+      end
+    elseif (0 <= f_max) then -- Case B
+      -- Show p10 .. p90 + headroom
+      local f_lo, f_hi = self:quantiles(0.1, 0.9)
+      local iqr = (f_hi - f_lo)/8*2
+      return f_lo - iqr, f_hi + iqr
+    else -- Case C
+      -- like A but reverse signs
+      local f_lo = self:quantiles(0.1)
+      if (f_min > 4*f_md) then
+        -- min is not unreasonably low: use it
+        return f_min, 0
+      else
+        -- put f_lo at 25% of (negative) y-scale
+        return f_lo / 0.75, 0
+      end
+    end
+  end
+end
+
+--------------------------------------------------------------------------------
+--- C struct based Histogram implementation
+Histogram_C.__index = Histogram_C
+
+function Histogram_C:__tostring()
+  local buf = string_builder:new()
+  local total = 0
+  for _, bin, count in self:funiter() do
+    buf:appendf("%.2g:%d", bin, count)
+    total = total + count
+  end
+  return string.format("Hist[%d]{%s}", total, buf:build(", "))
+end
+
+function Histogram_C:new(map)
+  local c_hist = libhist.hist_alloc()
+  assert(c_hist, "libhist.hist_alloc() returned nil (out of memory?!)")
+  ffi.gc(c_hist, libhist.hist_free)
+  if map then
+    for value, count in pairs(map) do
+      libhist.hist_insert(c_hist, value, count)
+    end
+  end
+  return c_hist
+end
+
+function Histogram_C:from_c_struct(c_hist, free_hist_when_done)
+  if c_hist == nil then
+    return ffi.cast("histogram_t*", nil)
+  end
+  if free_hist_when_done then
+    ffi.gc(c_hist, libhist.hist_free)
+  end
+  return c_hist
+end
+
+function Histogram_C:from_b64_string(b64_encoded_histo, str_len)
+  local c_hist = libhist.hist_alloc();
+  assert(c_hist, "libhist.hist_alloc() returned nil (out of memory?!)")
+  ffi.gc(c_hist, libhist.hist_free)
+
+  if str_len == nil then
+    str_len = string.len(b64_encoded_histo)
+  end
+
+  libhist.hist_deserialize_b64(c_hist, b64_encoded_histo, str_len)
+  return c_hist
+end
+
+function Histogram_C:bucket_count()
+  return libhist.hist_bucket_count(self)
+end
+
+function Histogram_C:clear()
+  libhist.hist_clear(self)
+end
+
+function Histogram_C:iter()
+  return coroutine.wrap(function()
+      for bin=0, self:bucket_count()-1 do
+        local value, count = self:bucket_idx(bin)
+        if count > 0 then
+          coroutine.yield(value, count)
+        end
+      end
+  end)
+end
+
+local function histogram_C_funiter_next(param, state)
+  local idx = state
+  if idx > param.limit then
+    return nil -- end iteration
+  end
+  local val, count = param.self:bucket_idx(idx)
+  return idx + 1, val, count
+end
+local function histogram_filter_count_zero(val, count)
+  return not (count == 0)
+end
+function Histogram_C:funiter()
```

Comment:
This is better than the coro-iteration, but depends on the fun.lua.

Iterating over the histogram with:
```
for i=1, h:bucket_count() do
  val, count = h:bucket_idx(idx)
  ...
end
```
should be good enough, no?

Consumers are free to wrap it into a fun iterator themselves.